### ISSUE TITLE
problem when running accumulated batchsize with MultistepAdamOptimizer

### ISSUE THREAD

**Issue by tobyyouup** [tensorflow/tensor2tensor#923]
### Description
when I run the accumulate batchsize with MultistepAdamOptimizer with step=8, on WMT en-fr translation model. The loss is very high, the loss should be below 10, but the loss is more than 500 with MultistepAdamOptimizer. 
Have you ever verify this code, have you met similar problems?
 

I use tensor2tensor 1.2.9, and tensorflow 1.4.0, So I modify the tensor2tensor/utils/multistep_optimizer.py file to my code, with some slightly changes on this file due to the tensorlfow API changes.

### **loss logs:**
INFO:tensorflow:loss = 10.122, step = 1
INFO:tensorflow:loss = 6.18374, step = 101 (46.639 sec)
INFO:tensorflow:loss = 6.14092, step = 201 (41.133 sec)
INFO:tensorflow:loss = 9.47875, step = 2
INFO:tensorflow:loss = 6.19239, step = 102 (189.436 sec)
INFO:tensorflow:loss = 5.58047, step = 202 (41.086 sec)
INFO:tensorflow:loss = 5.66014, step = 302 (41.950 sec)
INFO:tensorflow:loss = 5.44744, step = 402 (41.807 sec)
INFO:tensorflow:loss = 5.13551, step = 502 (41.701 sec)
INFO:tensorflow:loss = 5.42552, step = 602 (41.885 sec)
INFO:tensorflow:loss = 5.3761, step = 702 (41.941 sec)
INFO:tensorflow:loss = 5.47784, step = 802 (41.670 sec)
INFO:tensorflow:loss = 5.34478, step = 902 (41.726 sec)
INFO:tensorflow:loss = 5.33199, step = 1002 (41.854 sec)
INFO:tensorflow:loss = 5.74657, step = 1102 (41.652 sec)
INFO:tensorflow:loss = 5.50258, step = 1202 (41.995 sec)
INFO:tensorflow:loss = 5.48534, step = 1302 (41.915 sec)
INFO:tensorflow:loss = 5.92809, step = 1402 (41.691 sec)
INFO:tensorflow:loss = 4.60018, step = 1502 (41.739 sec)
INFO:tensorflow:loss = 6.00927, step = 1602 (42.109 sec)
INFO:tensorflow:loss = 5.58784, step = 1702 (41.685 sec)
INFO:tensorflow:loss = 5.97269, step = 1802 (41.991 sec)
INFO:tensorflow:loss = 6.01503, step = 1902 (41.689 sec)
INFO:tensorflow:loss = 5.6937, step = 2002 (42.356 sec)
INFO:tensorflow:loss = 9.19391, step = 2102 (41.657 sec)
INFO:tensorflow:loss = 9.24044, step = 2202 (41.596 sec)
INFO:tensorflow:loss = 10.6264, step = 2302 (41.812 sec)
INFO:tensorflow:loss = 9.32772, step = 2402 (41.886 sec)
INFO:tensorflow:loss = 10.9151, step = 2502 (41.556 sec)
INFO:tensorflow:loss = 7.1575, step = 2602 (41.598 sec)
INFO:tensorflow:loss = 9.30962, step = 2702 (44.236 sec)
INFO:tensorflow:loss = 10.0476, step = 2802 (41.533 sec)
INFO:tensorflow:loss = 22.3421, step = 2902 (41.369 sec)
INFO:tensorflow:loss = 26.0195, step = 3002 (41.660 sec)
INFO:tensorflow:loss = 36.6381, step = 3102 (41.527 sec)
INFO:tensorflow:loss = 60.19, step = 3202 (41.466 sec)
INFO:tensorflow:loss = 72.2689, step = 3302 (41.597 sec)
INFO:tensorflow:loss = 57.5796, step = 3402 (41.501 sec)
INFO:tensorflow:loss = 71.8269, step = 3502 (42.120 sec)
INFO:tensorflow:loss = 78.6678, step = 3602 (41.606 sec)
INFO:tensorflow:loss = 65.2495, step = 3702 (41.199 sec)
INFO:tensorflow:loss = 56.8085, step = 3802 (41.214 sec)
INFO:tensorflow:loss = 98.4963, step = 3902 (41.407 sec)
INFO:tensorflow:loss = 89.1331, step = 4002 (45.393 sec)
INFO:tensorflow:loss = 74.6041, step = 4102 (41.784 sec)
INFO:tensorflow:loss = 101.833, step = 4202 (41.613 sec)
INFO:tensorflow:loss = 96.3843, step = 4302 (41.733 sec)
INFO:tensorflow:loss = 118.482, step = 4402 (41.390 sec)
INFO:tensorflow:loss = 99.5724, step = 4502 (41.503 sec)
INFO:tensorflow:loss = 131.074, step = 4602 (41.688 sec)
INFO:tensorflow:loss = 128.016, step = 4702 (41.471 sec)
INFO:tensorflow:loss = 135.102, step = 4802 (41.399 sec)
INFO:tensorflow:loss = 138.732, step = 4902 (41.399 sec)
INFO:tensorflow:loss = 184.302, step = 5002 (41.510 sec)
INFO:tensorflow:loss = 173.179, step = 5102 (41.106 sec)
INFO:tensorflow:loss = 177.313, step = 5202 (40.863 sec)
INFO:tensorflow:loss = 197.261, step = 5302 (41.113 sec)
INFO:tensorflow:loss = 225.013, step = 5402 (41.227 sec)
INFO:tensorflow:loss = 220.586, step = 5502 (40.783 sec)
INFO:tensorflow:loss = 241.493, step = 5602 (40.931 sec)
INFO:tensorflow:loss = 267.513, step = 5702 (45.297 sec)
INFO:tensorflow:loss = 235.015, step = 5802 (40.712 sec)
INFO:tensorflow:loss = 236.259, step = 5902 (41.209 sec)
INFO:tensorflow:loss = 283.957, step = 6002 (41.336 sec)
INFO:tensorflow:loss = 273.576, step = 6102 (40.753 sec)
INFO:tensorflow:loss = 224.212, step = 6202 (40.882 sec)
INFO:tensorflow:loss = 233.976, step = 6302 (40.972 sec)
INFO:tensorflow:loss = 259.749, step = 6402 (41.279 sec)
INFO:tensorflow:loss = 244.262, step = 6502 (41.030 sec)
INFO:tensorflow:loss = 305.853, step = 6602 (41.331 sec)
INFO:tensorflow:loss = 353.021, step = 6702 (40.948 sec)
INFO:tensorflow:loss = 244.931, step = 6802 (41.089 sec)
INFO:tensorflow:loss = 368.809, step = 6902 (40.906 sec)
INFO:tensorflow:loss = 392.986, step = 7002 (41.300 sec)
INFO:tensorflow:loss = 396.258, step = 7102 (41.097 sec)
INFO:tensorflow:loss = 401.663, step = 7202 (41.055 sec)
INFO:tensorflow:loss = 408.524, step = 7302 (40.875 sec)
INFO:tensorflow:loss = 358.088, step = 7402 (41.063 sec)
INFO:tensorflow:loss = 380.768, step = 7502 (40.506 sec)
INFO:tensorflow:loss = 378.887, step = 7602 (40.914 sec)
INFO:tensorflow:loss = 400.1, step = 7702 (41.023 sec)
INFO:tensorflow:loss = 492.419, step = 7802 (41.079 sec)
INFO:tensorflow:loss = 460.063, step = 7902 (40.871 sec)
INFO:tensorflow:loss = 452.072, step = 8002 (41.183 sec)
INFO:tensorflow:loss = 454.302, step = 8102 (41.120 sec)
INFO:tensorflow:loss = 389.908, step = 8202 (40.899 sec)
INFO:tensorflow:loss = 613.68, step = 8302 (41.326 sec)

### **My tensor2tensor/utils/multistep_optimizer.py file:**

class MultistepAdamOptimizer(tf.train.AdamOptimizer):
  """Adam with SGD updates every n steps with accumulated gradients."""

  def __init__(self, learning_rate=0.001, beta1=0.9, beta2=0.999, epsilon=1e-8,
               use_locking=False, name="Adam", n=1):
    super(MultistepAdamOptimizer, self).__init__(
        learning_rate=learning_rate, beta1=beta1, beta2=beta2, epsilon=epsilon,
        use_locking=use_locking, name=name)
    self._n = n  # Call Adam optimizer every n batches with accumulated grads
    self._n_t = None  # n as tensor
    tf.logging.info("MultistepAdamOptimizer with step: %d", self._n)

  def _create_slots(self, var_list):
    """Create slot variables for Adam with accumulated gradients."""
    super(MultistepAdamOptimizer, self)._create_slots(var_list)
    first_var = min(var_list, key=lambda x: x.name)

    iter_var = tf.get_variable(name="iter", shape=[], dtype=tf.int32, initializer=tf.zeros_initializer,  trainable=False)
    
    for v in var_list:
      self._zeros_slot(v, "grad_acc", self._name)

  def _get_iter_variable(self):
    tf.get_variable_scope().reuse_variables()    
    return tf.get_variable(name="iter", shape=[], dtype=tf.int32)

  def _prepare(self):
    super(MultistepAdamOptimizer, self)._prepare()
    self._n_t = tf.convert_to_tensor(self._n, name="n")

  def _apply_cond(self, apply_fn, grad, var, *args, **kwargs):
    """Apply conditionally if counter is zero."""
    grad_acc = self.get_slot(var, "grad_acc")

    def apply_adam(grad_acc, apply_fn, grad, var, *args, **kwargs):
      total_grad = (grad_acc + grad) / tf.cast(self._n_t, grad.dtype)
      adam_op = apply_fn(total_grad, var, *args, **kwargs)
      with tf.control_dependencies([adam_op]):
        grad_acc_to_zero_op = grad_acc.assign(tf.zeros_like(grad_acc),
                                              use_locking=self._use_locking)
      return tf.group(adam_op, grad_acc_to_zero_op)

    def accumulate_gradient(grad_acc, grad):
      assign_op = tf.assign_add(grad_acc, grad, use_locking=self._use_locking)
      return tf.group(assign_op)  # Strip return value

    return tf.cond(
        tf.equal(self._get_iter_variable(), 0),
        lambda: apply_adam(grad_acc, apply_fn, grad, var, *args, **kwargs),
        lambda: accumulate_gradient(grad_acc, grad))

  def _apply_dense(self, grad, var):
    return self._apply_cond(
        super(MultistepAdamOptimizer, self)._apply_dense, grad, var)

  def _resource_apply_dense(self, grad, var):
    return self._apply_cond(
        super(MultistepAdamOptimizer, self)._resource_apply_dense, grad, var)

  def _apply_sparse_shared(self, grad, var, indices, scatter_add):
    return self._apply_cond(
        super(MultistepAdamOptimizer, self)._apply_sparse_shared, grad, var,
        indices, scatter_add)

  def _apply_sparse(self, grad, var):
    # TODO(fstahlberg): Implement a sparse version
    tf.logging.warning("MultistepAdamOptimizer does not support sparse updates")
    dense_grad = tf.convert_to_tensor(grad)
    return self._apply_cond(
        super(MultistepAdamOptimizer, self)._apply_dense, dense_grad, var)

  def _finish(self, update_ops, name_scope):
    """Updates beta_power variables every n batches and incrs counter."""
    iter_ = self._get_iter_variable()
   

    self._n_t = tf.Print(self._n_t, [iter_], summarize=10000, message="iter")
    beta1_power, beta2_power = self._get_beta_accumulators()
    with tf.control_dependencies(update_ops):
      with tf.colocate_with(iter_):

        def update_beta_op():
          update_beta1 = beta1_power.assign(
              beta1_power * self._beta1_t,
              use_locking=self._use_locking)
          update_beta2 = beta2_power.assign(
              beta2_power * self._beta2_t,
              use_locking=self._use_locking)
          return tf.group(update_beta1, update_beta2)
        maybe_update_beta = tf.cond(
            tf.equal(iter_, 0), update_beta_op, tf.no_op)
        with tf.control_dependencies([maybe_update_beta]):
          update_iter = iter_.assign(tf.mod(iter_ + 1, self._n_t),
                                     use_locking=self._use_locking)
    return tf.group(
        *update_ops + [update_iter, maybe_update_beta], name=name_scope)

I just change how to get the variable of "iter" in function  _create_slots(self, var_list) and 
_get_iter_variable(self)

Do you have any suggestions?

...

### Environment information
```
OS: Ununtu

$ pip freeze | grep tensor
# tensor2tensor==1.2.9
tensorflow-gpu==1.4.0
tensorflow-tensorboard==0.4.0rc3

$ python -V
#Python 3.6.3 :: Anaconda custom (64-bit)
```

**Comment by martinpopel** [comment]
I guess you have a strong reason for staying with the old T2T 1.2.9 (BTW: you could upgrade at least to T2T 1.4.2 and still stay [compatible](https://github.com/tensorflow/tensor2tensor/blob/v1.4.2/setup.py#L47) with TF 1.4), otherwise I suggest to update to the newest T2T.
My wild guess what is the cause of your problem:
Adam has a different optimal range of learning rates, compared to SGD and SGD+Momentum etc. T2T uses a multiplier constant of 500 (or 0.02) to make the optimal lr range of Adam and SGD similar. However, the place when this multiplier is applied has been changed in different T2T version. If you are mixing old code with new code, you may need to handle this as well.

**Comment by 282866787** [comment]
@tobyyouup Hi,friend, I have read your issue about distributed training , now I met some problem, can you give me some advice ? I use 1 master and 2 ps (both 8 gpus), but the training speed is 180 secs per 100 step, singel machine is just 55 secs ,and 1master 1 ps is 60 secs. I user sync training mode , I wonder if it is normal ?
